### PR TITLE
Scalar/histogram visualization via tensorboard

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

Scalar/histogram visualization via tensorboard

## How was this patch tested?

unit tests

## Related links or issues (optional)
#437 #438

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala [hunk 1]
```diff
@@ -44,6 +46,7 @@ object Train {
         conf.setAppName("Train Lenet on MNIST")
           .set("spark.akka.frameSize", 64.toString)
           .set("spark.task.maxFailures", "1")
+          .setMaster("local[1]")
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala]: What is this?!

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 2]
```diff
@@ -341,6 +354,41 @@ object DistriOptimizer {
     }
   }
 
+  private def saveSummary[T: ClassTag](
+        trainSummary: Option[TrainSummary],
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: remove the Option?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -341,6 +354,41 @@ object DistriOptimizer {
...
+        parameters: AllReduceParameter[T])(implicit ev: TensorNumeric[T]): Unit = {
+    val currentIteration = driverState[Int]("neval") - 1
+    if (trainSummary.isDefined && !trainSummary.get.getTrigger().isEmpty) {
+      val trigger = trainSummary.get.getTrigger()
+      val parametersTrigger = trigger.getOrElse[Trigger]("parameters", null)
+      if (null != parametersTrigger && parametersTrigger(driverState)) {
+        val model = getModelWithGradient(models, parameters)
+        val parametersTable = model.getParametersTable()
+        // Parallelize to create Histogram.
+        Engine.default.invoke(
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: How do you know it's done?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 4]
```diff
@@ -341,6 +354,41 @@ object DistriOptimizer {
...
   }
 
+  private def saveSummary[T: ClassTag](
+        trainSummary: Option[TrainSummary],
+        models: RDD[Cache[T]],
+        driverState: Table,
+        parameters: AllReduceParameter[T])(implicit ev: TensorNumeric[T]): Unit = {
+    val currentIteration = driverState[Int]("neval") - 1
+    if (trainSummary.isDefined && !trainSummary.get.getTrigger().isEmpty) {
+      val trigger = trainSummary.get.getTrigger()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: triggers? getTriggers?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 5]
```diff
@@ -341,6 +354,41 @@ object DistriOptimizer {
...
+        // Parallelize to create Histogram.
+        Engine.default.invoke(
+          parametersTable.keySet.toSeq.map(moduleName => () => {
+            val paramTable = parametersTable[Table](moduleName)
+            paramTable.keySet.foreach { paramName =>
+              trainSummary.get.addHistogram(
+                s"$moduleName/$paramName", paramTable[Tensor[T]](paramName), currentIteration)}
+          }))
+      }
+      val scalarTrigger = trigger.filter(!_._1.equals("parameters"))
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: scalarTriggers?

!_._1.equals("parameters") is hard to understand, provide a method with meaningful name?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 6]
```diff
@@ -497,6 +554,38 @@ object DistriOptimizer {
 
     trainedModel
   }
+
+  private def getModelWithGradient[T: ClassTag](
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: why not update the existing method?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+import java.util
+
+import com.intel.analytics.bigdl.optim.Trigger
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.visualization.tensorboard.{FileReader, FileWriter}
+
+import scala.reflect.ClassTag
+
+abstract class Summary(
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: You should add some doc for the class

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      logDir: String,
+      appName: String,
+      trigger: Map[String, Trigger]) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: triggers

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  protected val writer: FileWriter
+
+  def getTrigger(): Map[String, Trigger] = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: all public method need docs!!!

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    trigger
+  }
+
+  def addScalar(
+        tag: String,
+        value: Float,
+        step: Long): this.type = {
+    writer.addSummary(
+      com.intel.analytics.bigdl.utils.Summary.scalar(tag, value), step
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: why package full name?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        tag: String,
+        values: Tensor[T],
+        step: Long)(implicit ev: TensorNumeric[T]): this.type = {
+    writer.addSummary(
+      Summary.histogram[T](tag, values), step
+    )
+    this
+  }
+
+  def readScalar(tag: String): Array[(Long, Float, Double)]
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: what's it for?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    new ValidationSummary(logDir, appName, trigger)
+  }
+}
+
+// Logger for tensor board.
+object Summary {
+  // tag - specifying the type of the event, e.g. "learning rate", "loss", "throughput"
+  // scalar - a single float value, e.g. learning rate, loss, etc.
+  // returns a Summary protocol buffer containing a single scalar value.
+  // proto definition see  org.tensorflow/org.tensorflow/core/framework/summary.proto
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: why not javadoc?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+  def scalar(tag: String, scalar : Float): org.tensorflow.framework.Summary = {
+    val v = org.tensorflow.framework.Summary.Value.newBuilder().setTag(tag).setSimpleValue(scalar)
+    org.tensorflow.framework.Summary.newBuilder().addValue(v).build()
+  }
+
+  val limits = makeHistogramBuckets()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: private?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+  val counts = new Array[Int](limits.length)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: private or doc?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+  def histogram[T: ClassTag](
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: javadoc

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+        histogram.addBucket(counts(i))
+        histogram.addBucketLimit(limits(i))
+      }
+      i += 1
+    }
+    val v = org.tensorflow.framework.Summary.Value.newBuilder().setTag(tag).setHisto(histogram)
+    org.tensorflow.framework.Summary.newBuilder().addValue(v).build()
+  }
+
+  def bisectLeft(
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: private or doc?

File: pom.xml [hunk 17]
```diff
@@ -179,6 +179,7 @@
     <modules>
         <module>dl</module>
         <module>spark-version</module>
+        <module>visualization</module>
```
Comment (yiheng) [inline pom.xml]: change name to tensor-board?

File: visualization/pom.xml [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+                        <phase>verify</phase>
+                        <goals>
+                            <goal>jar-no-fork</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+
+            <plugin>
+                <groupId>org.scalastyle</groupId>
```
Comment (yiheng) [inline visualization/pom.xml]: why you need it?

File: visualization/src/main/java/org/tensorflow/framework/DataType.java [hunk 19]
```diff
@@ -0,0 +1,553 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (yiheng) [inline visualization/src/main/java/org/tensorflow/framework/DataType.java]: Can we generate this in build?

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+package com.intel.analytics.bigdl.visualization.tensorboard
+
+import java.io.File
+import java.net.InetAddress
+import java.util.concurrent.{LinkedBlockingDeque, TimeUnit}
+
+import org.tensorflow.util.Event
+
+class EventWriter(logDir: String, flushMillis: Int = 10000) extends Runnable {
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala]: doc!!!

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+  val eventQueue = new LinkedBlockingDeque[Event]()
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala]: private???

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+package com.intel.analytics.bigdl.visualization.tensorboard
+
+import java.util.concurrent.Executors
+
+import org.tensorflow.framework.Summary
+import org.tensorflow.util.Event
+
+class FileWriter(val logDirecotry : String, flushMillis: Int = 10000) {
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala]: Javadoc

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+  if (!logDir.exists()) logDir.mkdirs()
+  val eventWriter = new EventWriter(logDirecotry, flushMillis)
+  val pool = Executors.newFixedThreadPool(1)
+  pool.submit(eventWriter)
+  // adds a Summary protocol buffer to the event file.
+  // summary - a Summary protobuf String generted using TBLogger.scalar, etc.
+  // globalStep - a global count of the event, could be epoch, or iteration,
+  // as long as consistent.
+  // refer to https://github.com/dmlc/tensorboard/tree/master/python/tensorboard for simplified
+  // implementation of writers
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala]: javadoc style

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+package com.intel.analytics.bigdl.visualization.tensorboard
+
+import java.util.concurrent.Executors
+
+import org.tensorflow.framework.Summary
+import org.tensorflow.util.Event
+
+class FileWriter(val logDirecotry : String, flushMillis: Int = 10000) {
+  val logDir = new java.io.File(logDirecotry)
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala]: private if possible

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+  require(!logDir.exists() || logDir.isDirectory, s"FileWriter: can not create $logDir")
+  if (!logDir.exists()) logDir.mkdirs()
+  val eventWriter = new EventWriter(logDirecotry, flushMillis)
+  val pool = Executors.newFixedThreadPool(1)
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala]: make sure the thread is daemon or the program will not exit

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+      event.toBuilder.setWallTime(System.currentTimeMillis() / 1e3).setStep(globalStep).build())
+    this
+  }
+
+  def isEmpty(): Boolean = {
+    eventWriter.eventQueue.isEmpty
+  }
+
+  def close(): this.type = {
+    eventWriter.close()
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/FileWriter.scala]: why not close the pool

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala (hunk 21 above)
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/EventWriter.scala]: I think the queue should not be hold by EventWriter. It should be hold by FileWriter.

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/RecordWriter.scala [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+
+package com.intel.analytics.bigdl.visualization.tensorboard
+
+import java.io.{File, FileOutputStream}
+
+import com.google.common.primitives.{Ints, Longs}
+import netty.Crc32c
+import org.tensorflow.util.Event
+
+class RecordWriter(file: File) {
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/RecordWriter.scala]: doc, private!

File: visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/RecordWriter.scala [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+  def write(event: Event): Unit = {
+    val eventString = event.toByteArray
+    val header = Longs.toByteArray(eventString.length.toLong).reverse
+    outputStream.write(header)
+    outputStream.write(Ints.toByteArray(maskedCRC32(header).toInt).reverse)
+    outputStream.write(eventString)
+    outputStream.write(Ints.toByteArray(maskedCRC32(eventString).toInt).reverse)
+  }
+
+  def close(): Unit = {
```
Comment (yiheng) [inline visualization/src/main/scala/com/intel/analytics/bigdl/visualization/tensorboard/RecordWriter.scala]: all public method should have docs

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 3 above)
Comment (qiuxin2012) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: I don't know, change to invokeAndWait?

File: dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala [hunk 29]
```diff
@@ -0,0 +1,287 @@
...
+   */
+  def getSummaryTrigger(tag: String): Option[Trigger] = {
+    if (triggers.contains(tag)) {
+      Some(triggers(tag))
+    } else {
+      None
+    }
+  }
+
+  private[bigdl] def getScalarTriggers(): mutable.HashMap[String, Trigger] = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/utils/Summary.scala]: why return Map?

**Pull request merged**